FirebaseFunctionsSwift/Sources/Codable/Callable+Codable.swift

```diff
@@ -0,0 +1,184 @@
+  }
+}
+
+/**
+ * A `Callable` is reference to a particular Callable HTTPS trigger in Cloud Functions.
+ */
+public struct Callable<Request: Encodable, Response: Decodable> {
+  /**
+   * The timeout to use when calling the function. Defaults to 60 seconds.
+   */
+  public var timeoutInterval: TimeInterval {
+    get {
+      callable.timeoutInterval
+    }
+    set {
+      callable.timeoutInterval = newValue
+    }
+  }
+
+  enum CallableError: Error {
+    case internalError
+  }
+
+  private let callable: HTTPSCallable
+  private let encoder: StructureEncoder
+  private let decoder: StructureDecoder
+
+  init(callable: HTTPSCallable, encoder: StructureEncoder, decoder: StructureDecoder) {
+    self.callable = callable
+    self.encoder = encoder
+    self.decoder = decoder
+  }
+
+  /**
+   * Executes this Callable HTTPS trigger asynchronously.
+   *
+   * The data passed into the trigger must be of the generic `Request` type:
+   *
+   * The request to the Cloud Functions backend made by this method automatically includes a
+   * Firebase Instance ID token to identify the app instance. If a user is logged in with Firebase
+   * Auth, an auth ID token for the user is also automatically included.
+   *
+   * Firebase Instance ID sends data to the Firebase backend periodically to collect information
+   * regarding the app instance. To stop this, see `[FIRInstanceID deleteIDWithHandler:]`. It
+   * resumes with a new Instance ID the next time you call this method.
+   *
+   * - Parameter data: Parameters to pass to the trigger.
+   * - Parameter completion: The block to call when the HTTPS request has completed.
+   *
+   * - Throws: An error if any value throws an error during encoding.
+   */
+  public func call(_ data: Request,
+                   completion: @escaping (Result<Response, Error>)
+                     -> Void) {
+    do {
+      let encoded = try encoder.encode(data)
+
+      callable.call(encoded) { result, error in
+        do {
+          if let result = result {
+            let decoded = try decoder.decode(Response.self, from: result.data)
+            completion(.success(decoded))
+          } else if let error = error {
+            completion(.failure(error))
+          } else {
+            completion(.failure(CallableError.internalError))
+          }
+        } catch {
+          completion(.failure(error))
+        }
+      }
+    } catch {
+      completion(.failure(error))
+    }
+  }
+
+  /// Creates a directly callable function.
+  ///
+  /// This allows users to call a HTTPS Callable Funciton like a normal Swift function:
+  ///
+  ///     let greeter = functions.httpsCallable("greeter",
+  ///                                           requestType: GreetingRequest.self,
+  ///                                           responseType: GreetingResponse.self)
+  ///     try greeter(data) { result in
+  ///       print(result.greeting)
+  ///     }
```

Comment:
Can we use DocC formatting here? https://developer.apple.com/documentation/xcode/formatting-your-documentation-content#Add-Code-Listings

Reply:
Done. PTAL since I'm not sure how to test.